### PR TITLE
Add unsafe String API discussion to String manifesto

### PR DESCRIPTION
**Pull request by tardieu**
Proposal for an unsafe String API.

Unchecked, unvalidated access to the String buffer can sometimes enable significant performance gains compared to safe String operations.

swift-evolution discussion:
https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20170123/030773.html

Motivating example:
https://gist.github.com/tardieu/b6a9c4d53d56d089c58089ba8f6274b5

Tentative implementation in Swift 3:
https://gist.github.com/tardieu/7ca43d19b6197033dc39b138ba0e500e

### CODE REVIEW

File: docs/StringManifesto.md [hunk 1]
```diff
@@ -1124,6 +1124,55 @@ collection of `UInt32`, where values less than `0x80` represent the
...
+The following extension makes it possible to access and mutate the content of the String viewed as a contiguous buffer. 
+
+```swift
+extension String {
+    /// The base address of the continuous buffer containing the String elements
+    /// if the String is backed by a contiguous buffer.
+    ///
+    /// There is no guarantee that the String is backed by a contiguous buffer.
+    /// Even if it is, there is no guarantee that the buffer is writable.
+    public var baseAddress: UnsafeRawPointer? { get }
```
Comment (jrose-apple) [inline docs/StringManifesto.md]: Without jumping into the whole discussion, this should be an `UnsafeRawBufferPointer?` instead.

Comment (tardieu) [inline docs/StringManifesto.md]: Sure it could be. I decided to mimic the `StringBuffer` code (which predates `UnsafeRawBufferPointer`). My main motivation for not going for an `UnsafeRawBufferPointer` is the possible mismatch between the buffer size measured in bytes and measured in elements. I would expect the user to derive either an `UnsafeBufferPointer<UInt8>` (or `CChar` or raw) or an `UnsafeBufferPointer<UInt16>` from the base address. I am concerned that baking in the choice of width into the type even indirectly as an `UnsafeRawBufferPointer` may be a source of confusion.

**Review by airspeedswift — COMMENTED** [review]
I agree with the principle of adding explicit mention of unsafe APIs to the manifesto. This PR goes a little further than I'd want for a manifesto though – explicit API design should be left for the proposal stage.

It's also likely that we'll want some form of unification with the `withUnsafe` methods on `Array` where possible.

File: docs/StringManifesto.md (hunk 1 above)
Comment (airspeedswift) [inline docs/StringManifesto.md]: I don't think it's appropriate to put a `baseAddress` method directly on `String` for a few reasons. Primarily, it encourages extracting the pointer and then using it in code that may cause the string to reallocate. The `withUnsafe` closure-based API is styled to help avoid this.

**Review by airspeedswift — CHANGES_REQUESTED** [review]

**Comment by karwa** [comment]
This kind of contiguous raw access is likely to become codified as a protocol soon (see: https://bugs.swift.org/browse/SR-3631), so String could simply conform to it.

**Comment by tardieu** [comment]
I agree a `withUnsafe` closure-based API is cleaner in principle but I am concerned with performance. I have implemented such an API here: https://gist.github.com/tardieu/2ef5393a3037cfa639be7032193475c3.

On my macbook, the performance penalty of `withUnsafe` is significant.
```
$ swiftc -O -whole-module-optimization withUnsafe.swift
$ ./withUnsafe
closure: 179.966374ms
pointer: 111.830017ms
closure: 174.908947ms
pointer: 119.457985ms
closure: 172.952545ms
pointer: 113.70196ms
```

The penalty of course depend on the cost of the operations performed on the String. In this example, I use: `ptr.pointee = 65 + UInt8(i % 26)` to prevent the compiler from vectorizing the pointer-based loop and have a "fair" comparison. But as a result, the performance gap between the two versions is reduced compared to the expected typical use case of simply reading or writing a a few bytes.

I annotated `withUnsafe` with `@inline(__always)` but I suppose the standard library could still provide a faster implementation leveraging transparent inlining. If performance parity can be achieved then great... otherwise I would argue that performance should remain the deciding factor here.

It is of course possible to abuse a `withUnsafe` API and leak the pointer but then the validity of the pointer is unspecified, i.e., it should not be done. On the other hand, a `baseAddress` kind of API can specify sufficient conditions for the pointer to remain valid.

What is the proper way to prototype `withUnsafe` with maximum performance? Extend and rebuild the standard library?

**Comment by jckarter** [comment]
**Correctness** must supersede performance. We can easily improve the optimizer; it's much harder to backpedal broken API. You should file a bug to fix the optimizer.

**Comment by tardieu** [comment]
I agree that a `withUnsafe[Mutable]BufferPointer` API is better than nothing and I can revise this pull request to only discuss these two methods for now. I don't think however that we will be able to replace our C code for String processing with Swift code if performance continues to be a secondary concern.

**Comment by gottesmm** [comment]
> On Feb 21, 2017, at 12:20 PM, Olivier Tardieu <notifications@github.com> wrote:
> 
> I agree that a withUnsafe[Mutable]BufferPointer API is better than nothing and I can revise this pull request to only discuss these two methods for now. I don't think however that we will be able to replace our C code for String processing with Swift code if performance continues to be a secondary concern.
> 
> 

Question, is the closure not being inlined? In such a case, it should be specialized. If it is not being specialized that may be due to generics. Can you confirm?

Michael

**Comment by tardieu** [comment]
The the `baseAddress` vs. `withUnsafe` discussion does not address other key elements of this pull request: element count and capacity.

While it makes sense to align the String and Array APIs, the current Array API is much more limited than this pull request. Today for instance, the following array code crashes:
```
var array = [0,1,2]
array.reserveCapacity(10)
array.withUnsafeMutableBufferPointer { $0[3] = 3 }
```
Only array.count elements are exposed via the buffer pointer, not the full capacity of the array. Moreover the element count is a read-only property.

In order for a closure-based API to permit shrinking or extending a String (up to its preallocated capacity), it needs to pass into the closure something more expressive than an UnsafeMutableBufferPointer. BTW, the same is true for arrays of course.

To a lesser degree element width is also a concern and a departure from the Array API. The `withUnsafeMutableBufferPointer` on Array is generic in the element type. Strings are generic in a way that is not reflected by the type system. This has consequences for the type of buffer pointers and capacity management. For instance, the current description of the `reserveCapacity` function on String is rather vague—Reserves enough space in the string’s underlying storage to store the specified number of ASCII characters—with no discussion of element width.

In short, there is more to the current pull request than a fixed-size contiguous memory buffer and I'd like to see a discussion of these other capabilities.

**Comment by tardieu** [comment]
I believe the closure is correctly inlined by the compiler. However, there is more to the `withUnsafe` method that the closure application. The "superior" correctness of the closure form relies on the fact that the baseAddress is obtained on every invocation so that it up to date (`self._core.nativeBuffer!.start`). It will be slower unless the optimizer is able to prove the value is constant across invocations and reuse it. From the look of the assembly code, I believe this is not the case.

This is a typical situation where I would rather have pinning than heroic compiler optimization. Pinning guarantees both safety and performance. I understand a closure-based API ties together pinning and lifetime in a rather foolproof way (assuming unique owner) but I would argue... again... than unsafe APIs are not for fools anyway. I would be happy to explicitly pin/unpin the string buffer in order to obtain a pointer to it. 

The Array `withUnsafeMutableBufferPointer` API already has this warning (in red):
"Do not rely on anything about the array that is the target of this method during execution of the body closure; it might not appear to have its correct value. Instead, use only the UnsafeMutableBufferPointer argument to body." So `withUnsafeMutableBufferPointer` is only so safe.

File: docs/StringManifesto.md (hunk 1 above)
Comment (tardieu) [inline docs/StringManifesto.md]: I have dropped the proposed String extension from this pull request and replaced it with a reference to the Array API.

**Comment by jckarter** [comment]
@tardieu Are you not able to hoist the `withUnsafe` call so that the closure encloses all of the code that needs the pointer?

**Comment by jckarter** [comment]
@tardieu The Array APIs intentionally don't expose the unused capacity because, in their current design, there's no way to safely initialize that capacity and have the initialized elements be adopted by the collection. There would need to be an interface that let the closure specify the new count of valid elements at any point the closure exits. Strings would have the same problems, albeit with at least the not-inconsiderable simplification that they only ever contain POD elements so destroying elements isn't a concern.

**Review by tardieu — COMMENTED** [review]
I have dropped the proposed String extension from this pull request and replaced it with a reference to the Array API.

**Comment by tardieu** [comment]
@jckarter It is often possible to hoist the `withUnsafe` code but not always, in particular when using persistent string buffers inside of an API.

**Comment by karwa** [comment]
@tardieu inability to unsafely initialise Array elements is also a known issue. https://bugs.swift.org/browse/SR-3087

You could imagine some other protocol which refined `ContiguouslyStored` and which allowed you to mutate a contiguous region of a given size, promising that at the end of your mutation all those elements will be filled. For example:

```swift
protocol UnsafelyInitialisable: ContiguouslyStored {
     // - The given buffer pointer promises to have capacity for at least `count` elements
     // - The `perform` closure promises that `count` elements will be initialised when it returns
     mutating func withUnsafeMutableBufferPointer<T>(count: Int, perform: (UnsafeMutableBufferPointer<Element>)->T) -> T
}
```

**Comment by jckarter** [comment]
The `with`-bracket is especially important for vending mutating pointers, since the value semantics model relies on single ownership of mutable memory. Taking a returned `baseAddress` and trying to mutate through it will always be UB in our current model. We'd need to leverage proper ownership tracking to make that kind of thing safe (and at the point we're able to do that, I totally agree that's a better interface than `with` blocks).

**Comment by tardieu** [comment]
@airspeedswift I have dropped the proposed String extension from this pull request and replaced it with a reference to the Array API. Is there anything else I need to do here?

**Comment by dabrahams** [comment]
I don't expect to have exactly those APIs on `String`, FWIW.  We don't want to construct an array of `UInt16`s when the string stores Latin-1, for example.

**Comment by airspeedswift** [comment]
I think this probably would merit from more discussion on the list before making it into the manifesto via a PR. As Dave points out, the `withUnsafe` pattern from `Array` doesn't necessarily map directly onto `String` due to the different underlying representations that might be used.

**Comment by CodaFi** [comment]
This pull request is 2 years old and there hasn't been a proper discussion of this on swift evolution as far as I can see. I'm going to close this.  @tardieu, if you wish to pursue this in Swift 5, please reopen this discussion on the forums and send us an updated pull request with the results.

Thanks!
